Snowflake RSA connection doesn't work with Hostname setting

### Describe the bug

Snowflake RSA connection doesn't work with Hostname setting

### To Reproduce

Just try to connect with the RSA key but instead of Account Name toggle to Host Name

<img width="1098" alt="Image" src="https://github.com/user-attachments/assets/ede41d8a-6102-4afd-8eb5-2ccab5ce6896" />

You will get an error. Doesn't matter what you enter in that section even a simple character

<img width="972" alt="Image" src="https://github.com/user-attachments/assets/07250236-8c3f-4904-a233-1ae973203343" />

### Expected behavior

If we don't support hostname with RSA key then remove the option

### Logs

`POST /api/database 400 205ms (0 DB calls) {:metabase-user-id 1} 
{:message "Cannot invoke \"java.lang.CharSequence.length()\" because \"this.text\" is null"}`

### Information about your Metabase installation

```JSON
53.1 and master
```

### Severity

Might be confusing for someone who doesn't know what is going on

### Additional context

_No response_

same thing happens when connecting to MongoDB via connection string

```
2025-04-12 13:32:39,376 ERROR driver.util :: Failed to connect to Database
java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
	at java.base/java.util.regex.Matcher.getTextLength(Unknown Source)
	at java.base/java.util.regex.Matcher.reset(Unknown Source)
	at java.base/java.util.regex.Matcher.<init>(Unknown Source)
	at java.base/java.util.regex.Pattern.matcher(Unknown Source)
	at clojure.core$re_matcher.invokeStatic(core.clj:4909)
	at clojure.core$re_find.invokeStatic(core.clj:4951)
	at clojure.core$re_find.invoke(core.clj:4951)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invokeStatic(ssh.clj:114)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invoke(ssh.clj:109)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:158)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:154)
	at metabase.driver.mongo.connection$do_with_mongo_client.invokeStatic(connection.clj:89)
	at metabase.driver.mongo.connection$do_with_mongo_client.invoke(connection.clj:85)
	at metabase.driver.mongo$fn__131236.invokeStatic(mongo.clj:53)
	at metabase.driver.mongo$fn__131236.invoke(mongo.clj:51)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__55450.invoke(util.clj:173)
	at clojure.core$binding_conveyor_fn$fn__5842.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2025-04-12 13:32:39,382 ERROR driver.util :: Failed to connect to Database
java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
	at java.base/java.util.regex.Matcher.getTextLength(Unknown Source)
	at java.base/java.util.regex.Matcher.reset(Unknown Source)
	at java.base/java.util.regex.Matcher.<init>(Unknown Source)
	at java.base/java.util.regex.Pattern.matcher(Unknown Source)
	at clojure.core$re_matcher.invokeStatic(core.clj:4909)
	at clojure.core$re_find.invokeStatic(core.clj:4951)
	at clojure.core$re_find.invoke(core.clj:4951)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invokeStatic(ssh.clj:114)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invoke(ssh.clj:109)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:158)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:154)
	at metabase.driver.mongo.connection$do_with_mongo_client.invokeStatic(connection.clj:89)
	at metabase.driver.mongo.connection$do_with_mongo_client.invoke(connection.clj:85)
	at metabase.driver.mongo$fn__131236.invokeStatic(mongo.clj:53)
	at metabase.driver.mongo$fn__131236.invoke(mongo.clj:51)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__55450.invoke(util.clj:173)
	at clojure.core$binding_conveyor_fn$fn__5842.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2025-04-12 13:32:39,386 ERROR api.database :: Cannot connect to Database
clojure.lang.ExceptionInfo: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null {:message "Cannot invoke \"java.lang.CharSequence.length()\" because \"this.text\" is null"}
	at metabase.driver.util$can_connect_with_details_QMARK_.invokeStatic(util.clj:190)
	at metabase.driver.util$can_connect_with_details_QMARK_.doInvoke(util.clj:161)
	at clojure.lang.RestFn.invoke(RestFn.java:445)
	at metabase.api.database$test_database_connection.invokeStatic(database.clj:731)
	at metabase.api.database$test_database_connection.doInvoke(database.clj:721)
	at clojure.lang.RestFn.invoke(RestFn.java:428)
	at metabase.api.database$test_connection_details106942__106943.invokeStatic(database.clj:785)
	at metabase.api.database$test_connection_details106942__106943.invoke(database.clj:765)
	at metabase.api.database$fn__106950$fn__106956.invoke(database.clj:809)
	at metabase.api.macros$_core_fn93586__93587$core_fn__93588.invoke(macros.clj:405)
	at metabase.api.macros$endpoint_handler_STAR_93630__93632$handler__93636.invoke(macros.clj:601)
	at metabase.api.macros$build_ns_handler93663__93664$ns_handler_STAR___93665.invoke(macros.clj:665)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.server.middleware.auth$enforce_authentication$fn__62465.invoke(auth.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103496.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433$f__103434$respond_SINGLEQUOTE___103435.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__103496.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433.invoke(core.clj:200)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.server.routes$fn__112496$fn__112497.invoke(routes.clj:70)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433.invoke(core.clj:200)
	at compojure.core$make_context$handler__103461.invoke(core.clj:290)
	at compojure.core$make_context$fn__103465.invoke(core.clj:300)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433$f__103434$respond_SINGLEQUOTE___103435.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103414.invoke(core.clj:153)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433$f__103434$respond_SINGLEQUOTE___103435.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103414.invoke(core.clj:153)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433$f__103434$respond_SINGLEQUOTE___103435.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103414.invoke(core.clj:153)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433$f__103434$respond_SINGLEQUOTE___103435.invoke(core.clj:197)
	at compojure.core$wrap_route_matches$fn__103414.invoke(core.clj:153)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433$f__103434$respond_SINGLEQUOTE___103435.invoke(core.clj:197)
	at metabase.api.util.handlers$_route_map_handler$fn__103496.invoke(handlers.clj:24)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at metabase.api.util.handlers$_route_map_handler$fn__103496.invoke(handlers.clj:23)
	at metabase.api.open_api.HandlerWithOpenAPISpec.invoke(open_api.clj:42)
	at clojure.lang.Var.invoke(Var.java:395)
	at compojure.core$routes$fn__103433$f__103434.invoke(core.clj:198)
	at compojure.core$routes$fn__103433.invoke(core.clj:200)
	at metabase.server.middleware.exceptions$catch_uncaught_exceptions$fn__62642.invoke(exceptions.clj:111)
	at metabase.server.middleware.exceptions$catch_api_exceptions$fn__62639.invoke(exceptions.clj:100)
	at metabase.server.middleware.log$log_api_call$fn__118736$fn__118737$fn__118738.invoke(log.clj:232)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invokeStatic(diagnostic.clj:17)
	at metabase.driver.sql_jdbc.execute.diagnostic$do_with_diagnostic_info.invoke(diagnostic.clj:12)
	at metabase.server.middleware.log$log_api_call$fn__118736$fn__118737.invoke(log.clj:223)
	at toucan2.execute$do_with_call_counts.invokeStatic(execute.clj:112)
	at toucan2.execute$do_with_call_counts.invoke(execute.clj:103)
	at metabase.server.middleware.log$log_api_call$fn__118736.invoke(log.clj:222)
	at metabase.server.middleware.browser_cookie$ensure_browser_id_cookie$fn__127366.invoke(browser_cookie.clj:40)
	at metabase.server.middleware.security$add_security_headers$fn__62598.invoke(security.clj:273)
	at metabase.server.middleware.json$wrap_json_body$fn__48850.invoke(json.clj:128)
	at metabase.server.middleware.offset_paging$handle_paging$fn__126164.invoke(offset_paging.clj:48)
	at metabase.server.middleware.json$wrap_streamed_json_response$fn__48838.invoke(json.clj:83)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__127633.invoke(keyword_params.clj:55)
	at ring.middleware.params$wrap_params$fn__127652.invoke(params.clj:77)
	at metabase.server.middleware.misc$maybe_set_site_url$fn__118154.invoke(misc.clj:59)
	at metabase.server.middleware.session$reset_session_timeout$fn__120940.invoke(session.clj:270)
	at metabase.server.middleware.session$bind_current_user$fn__120933$fn__120934.invoke(session.clj:245)
	at metabase.request.session$do_with_current_user.invokeStatic(session.clj:47)
	at metabase.request.session$do_with_current_user.invoke(session.clj:30)
	at metabase.server.middleware.session$bind_current_user$fn__120933.invoke(session.clj:244)
	at metabase.server.middleware.session$wrap_current_user_info$fn__120925.invoke(session.clj:204)
	at metabase.analytics.sdk$embedding_mw$embedding_mw_fn__65418.invoke(sdk.clj:79)
	at metabase.server.middleware.session$wrap_session_id$fn__120897.invoke(session.clj:82)
	at metabase.server.middleware.auth$wrap_static_api_key$fn__62473.invoke(auth.clj:38)
	at ring.middleware.cookies$wrap_cookies$fn__127553.invoke(cookies.clj:200)
	at metabase.server.middleware.misc$add_content_type$fn__118136.invoke(misc.clj:28)
	at metabase.server.middleware.misc$disable_streaming_buffering$fn__118162.invoke(misc.clj:75)
	at ring.middleware.gzip$wrap_gzip$fn__127595.invoke(gzip.clj:86)
	at metabase.server.middleware.request_id$wrap_request_id$fn__116664.invoke(request_id.clj:10)
	at metabase.server.middleware.misc$bind_request$fn__118165$fn__118166.invoke(misc.clj:87)
	at metabase.request.current$do_with_current_request.invokeStatic(current.clj:19)
	at metabase.request.current$do_with_current_request.invoke(current.clj:15)
	at metabase.server.middleware.misc$bind_request$fn__118165.invoke(misc.clj:86)
	at metabase.server.middleware.ssl$redirect_to_https_middleware$fn__127382.invoke(ssl.clj:51)
	at clojure.lang.Var.invoke(Var.java:395)
	at metabase.server.instance$async_proxy_handler$fn__62982.invoke(instance.clj:76)
	at metabase.server.instance.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1598)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:753)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:501)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:287)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:199)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
	at java.base/java.util.regex.Matcher.getTextLength(Unknown Source)
	at java.base/java.util.regex.Matcher.reset(Unknown Source)
	at java.base/java.util.regex.Matcher.<init>(Unknown Source)
	at java.base/java.util.regex.Pattern.matcher(Unknown Source)
	at clojure.core$re_matcher.invokeStatic(core.clj:4909)
	at clojure.core$re_find.invokeStatic(core.clj:4951)
	at clojure.core$re_find.invoke(core.clj:4951)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invokeStatic(ssh.clj:114)
	at metabase.util.ssh$include_ssh_tunnel_BANG_.invoke(ssh.clj:109)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:158)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:154)
	at metabase.driver.mongo.connection$do_with_mongo_client.invokeStatic(connection.clj:89)
	at metabase.driver.mongo.connection$do_with_mongo_client.invoke(connection.clj:85)
	at metabase.driver.mongo$fn__131236.invokeStatic(mongo.clj:53)
	at metabase.driver.mongo$fn__131236.invoke(mongo.clj:51)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__55450.invoke(util.clj:173)
	at clojure.core$binding_conveyor_fn$fn__5842.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
2025-04-12 13:32:39,392 DEBUG middleware.log :: POST /api/database 400 147ms (1 DB calls) {:metabase-user-id 1} 
{:message "Cannot invoke \"java.lang.CharSequence.length()\" because \"this.text\" is null"}
```

![Image](https://github.com/user-attachments/assets/b1083ca1-3d82-4c57-abdd-a92fc65bda4d)

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -87,15 +87,15 @@
 
 (defn- handle-conn-uri [details user account private-key-file]
   (let [existing-conn-uri (or (:connection-uri details)
+                              (when-let [sub (:subname details)]
+                                (format "jdbc:snowflake:%s" sub))
                               (format "jdbc:snowflake://%s.snowflakecomputing.com" account))
         opts-str (sql-jdbc.common/additional-opts->string :url
                                                           {:user (codec/url-encode user)
                                                            :private_key_file (codec/url-encode (.getCanonicalPath ^File private-key-file))})
         new-conn-uri (sql-jdbc.common/conn-str-with-additional-opts existing-conn-uri :url opts-str)]
     (-> details
-        (assoc :connection-uri new-conn-uri)
-        ;; The Snowflake driver uses the :account property, but we need to drop the region from it first
-        (assoc :account (first (str/split account #"\."))))))
+        (assoc :connection-uri new-conn-uri))))
 
 (defn- resolve-private-key
   "Convert the private-key secret properties into a private_key_file property in `details`.
@@ -194,12 +194,6 @@
     ;; https://support.snowflake.net/s/question/0D50Z00008WTOMCSA5/
     (-> (merge {:classname                                  "net.snowflake.client.jdbc.SnowflakeDriver"
                 :subprotocol                                "snowflake"
-                ;; see https://github.com/metabase/metabase/issues/22133
-                :subname                                    (let [base-url (if (and use-hostname (string? host) (not (str/blank? host)))
-                                                                             (cond-> host
-                                                                               (not= (last host) \/) (str "/"))
-                                                                             (str account ".snowflakecomputing.com/"))]
-                                                              (str "//" base-url))
                 :client_metadata_request_use_connection_ctx true
                 :ssl                                        true
                 ;; keep open connections open indefinitely instead of closing them. See #9674 and
@@ -213,6 +207,15 @@
                 ;; [[metabase.public-settings/start-of-week]] Setting.
                 :week_start                                 (start-of-week-setting->snowflake-offset)}
                (-> details
+                   ;; see https://github.com/metabase/metabase/issues/22133
+                   (update :subname (fn [subname]
+                                      (if subname
+                                        subname
+                                        (let [base-url (if (and use-hostname (string? host) (not (str/blank? host)))
+                                                         (cond-> host
+                                                           (not= (last host) \/) (str "/"))
+                                                         (str account ".snowflakecomputing.com/"))]
+                                          (str "//" base-url)))))
                    ;; original version of the Snowflake driver incorrectly used `dbname` in the details fields instead
                    ;; of `db`. If we run across `dbname`, correct our behavior
                    (set/rename-keys {:dbname :db})

--- a/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
+++ b/modules/drivers/snowflake/src/metabase/driver/snowflake.clj
@@ -95,7 +95,9 @@
                                                            :private_key_file (codec/url-encode (.getCanonicalPath ^File private-key-file))})
         new-conn-uri (sql-jdbc.common/conn-str-with-additional-opts existing-conn-uri :url opts-str)]
     (-> details
-        (assoc :connection-uri new-conn-uri))))
+        (assoc :connection-uri new-conn-uri)
+        ;; The Snowflake driver uses the :account property, but we need to drop the region from it first (#30376 comment)
+        (m/assoc-some :account (some-> account (str/split #"\.") first)))))
 
 (defn- resolve-private-key
   "Convert the private-key secret properties into a private_key_file property in `details`.

--- a/modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj
+++ b/modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj
@@ -488,7 +488,12 @@
              net.snowflake.client.jdbc.SnowflakeSQLException
              (can-connect? (assoc (:details (mt/db)) :db (mt/random-name))))
             "can-connect? should throw for Snowflake databases that don't exist (#9511)")
-
+        (is (can-connect? (-> (:details (mt/db))
+                              (assoc :host (str (get-in (mt/db) [:details :account])
+                                                ".snowflakecomputing.com")
+                                     :use-hostname true)
+                              (dissoc :account)))
+            "can-connect? with host and no account")
         (when (and pk-key pk-user)
           (mt/with-temp-file [pk-path]
             (mt/with-temp [:model/Secret {path-secret-id :id} {:name "Private key for Snowflake"
@@ -505,6 +510,13 @@
                                                                         :value (mt/bytes->base64-data-uri (u/string-to-bytes pk-key))}]
               (testing "private key authentication via uploaded keys or local key with path stored in a secret"
                 (spit pk-path pk-key)
+                (is (can-connect? (-> (:details (mt/db))
+                                      (assoc :host (str (get-in (mt/db) [:details :account])
+                                                        ".snowflakecomputing.com")
+                                             :use-hostname true)
+                                      (dissoc :password :account)
+                                      (merge {:db pk-db :user pk-user} {:private-key-id path-secret-id})))
+                    "can-connect? with pk, host and no account")
                 (doseq [to-merge [;; uploaded string
                                   {:private-key-value pk-key
                                    :private-key-options "uploaded"}

